### ISSUE TITLE
SEC 380 - Support for distributed Auth Cache in Liberty

### ISSUE THREAD

**Issue by atosak** [OpenLiberty/open-liberty#9005]
## Description of the high level feature, including any external spec links:  
<br/>
This feature will provide support for use of a shared authentication cache between Liberty servers. It will resolve a number of security issues related to single signon and the logout process. Today, when requests for a user are routed between 2 or more servers, should the user log out of one, the LTPA token is still valid in the other servers till it times out.

##
#### Before proceeding to any items below (active development), this feature must be prioritized on the backlog, and have either the "Design Approved" or "No Design Approved" labels.  Follow the Feature and UFO Approval Process.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: https://ibm.box.com/s/jj3emsdtoabq3dkw59ckbdiherljghzx
- FTS: https://github.com/openliberty/open-liberty/issues/19012
- Beta Blog Post (if applicable): https://github.com/OpenLiberty/open-liberty/issues/20209
- Blog Post: https://github.com/OpenLiberty/open-liberty/issues/22913
- STE: https://ibm.box.com/s/g8003aarjclc04e76pfl7qg7oovf9pst
- Guide Issue: https://github.com/OpenLiberty/guides-common/issues/849

Doc /ID Issues:
- https://github.com/OpenLiberty/docs/issues/5388
- https://github.com/OpenLiberty/docs/issues/5389
- https://github.com/OpenLiberty/docs/issues/5704

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [x] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the release branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [x] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [x] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [x] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by mbvreddy** [comment]
Lack of this feature make Single Page Applications handicapped without session affinity. Request that hit another server in cluster fail to do SSO authentication due to lack of Subject in that server's auth cache thereby forcing "redirect to OP" (in case of OIDC). In case of making calls using AJAX, it cannot handle the redirection to OP and force application to reload the complete webpage in this scenario affecting user experience with webpage.

**Comment by SteveFreeman** [comment]
We're seeing the underlying issue for this with multiple applications and we have no way to compensate right now -- so I vote for bumping this up the priority ladder.

**Comment by jtmulvey** [comment]
Adding link to Aha idea requesting this support https://cloud-platform.ideas.aha.io/ideas/TWAS-I-175

**Comment by garypicher** [comment]
UFO Review Comments:

Slide 28 (and throughout the config in general) - change "jCacheProvider" to "cacheProvider", and more generally, remove the "j" in front of any new "jCache*" configuration elements / properties

```
JVH: Updated.
```

Slide 29 - indicate that despite no Liberty default, there is a vendor default

```
JVH: Updated.
```
Slide 31 - is jCacheManagerRef required if the manager is nested?

```
JVH: It is not required if nested. I have added new slides an updated configuration examples to show using the nested configuration.
```

General comment:  Should it be explicitly stated somewhere early on that the distributed auth cache and the logged out token cache are configured separately?

```
JVH: The document shows configuring two separate caches. I have updated the descriptions for the REF attributes for both to say they should use UNIQUE cache references.
```

Slide 35 - Should there be a warning message if the cacheManagerRef is configured and so are the other values that we ignore when a cacheManagerRef is configured?

```
JVH: Updated slide to indicate a warning message will be issued.
```

Slide 38 - Typo, "loggedOutCookieCache" --> "loggedOutCookieCacheRef"

```
JVH: Updated.
```
Slide 38 - mpJwt is actually targeted for 22.0.0.1

```
JVH: Updated.
```

Slide 47 - Will performance testing be run to quantify how much improvement we actually see for certain key paths?

```
JVH: Added slide indicating that performance testing team will take a look and ensure we didn't degrade performance for the existing functionality.
```

Slide 48 - Anything we should do with our operators to make configuring this easier in the cloud?

```
JVH: I talked to @leochr about whether there was anything we should to make configuration easier in the cloud, and we came to the conclusion that there isn't anything at this time.
```

Slide 53 - "Deserialization errors" --> "Serialization errors"

```
JVH: Updated.
```

Slide 53 - Talk to Jared about concerns with different configurations for each server, do we need to account for errors that might occur in such a case?

```
JVH: Jared and I talked about this finding and decided that currently JCache throws exceptions when the requested cache does not match the existing cache, so we feel at this time there is nothing to be done here. In the future, it is possible we may add an internal cache if there is a need to communicate configuration / versions between Liberty servers.
```

General comment:  Update HTTP session cache documentation so that examples use JCache

**Comment by jvanhill** [comment]
As we have had some status requests, I would like to note that I am actively working on this feature. 

If you want to test it out, the 22.0.0.5 Open Liberty beta contains this functionality, and you can test it there (or any subsequent beta release).

- https://openliberty.io/start/#runtime_betas

Current major blockers to releasing to GA are some unintended issues with classloading as well some intermittent issues where the authentication cache is cleared on server startup.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

```
Yes, in the serviceability section of the UFO, which covers 3 slides. They have been addressed in the implementation.
```

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

```
a. Scenarios tested:
   - JCache provider not available at startup
   - Non-serializable class contained within the Subject
   - Caching provider class not found
   - Customer defined principal / credential stored in subject not found in the common library configured in JCache
   - JCache provider fails caching operation
b. Jackson Leonard (@jacksonleonard-zz)
c. Jackson agrees that the serviceability for the scenarios is sufficient to avoid PMRs and that L2 should be able to quickly address problems without need to engage L3.
```

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

```
a. Nichole Stewart (@nstewart0206)
b. Nichole agrees that the serviceability of the problems they encountered was sufficient to resolve the issues she encountered.
```

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

```
Authentication Cache and Logged Out Cookie Cache
 - WASL2: SEC
 - WAS: L3 Core Security
HTTP Session Cache
 - WASL2: WEB HTTP Sessions
 - WAS: L3 Sessions

These are the same queues that already cover these functionalities.
```

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

```
No JSON events are generated.
```

**Comment by gnadell** [comment]
@jvanhill 
WAS Security L2 has requested STE slides for this feature. The STE template can be found at the links below. You can use either one to create the education.
Slide Template: https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm
Github Template: https://pages.github.ibm.com/WASL3/site/STE/about
Please upload the completed slides to the same 'STE Archive' BOX folder and let @tngiang73 or me know they're complete or provide us the Github link.  

Thanks!

**Comment by chirp1** [comment]
@jvanhill Approving. Jesse Opened this ID issue in August and provided the info to document. David is writing the docs, and David and Jesse are working through reviews.
